rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,346 @@
+---
+feature: sign-commits
+start-date: 2021-08-14
+author: Las Safin
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues: https://github.com/NixOS/rfcs/pull/34
+---
+
+# Summary
+[summary]: #summary
+
+- Add a way for flake inputs that point to Git repositories to be verified for trust.
+- Speed up fetching Git repositories with Nix.
+- Recommend and eventually require that people who commit directly to
+  https://github.com/NixOS/nixpkgs and https://github.com/NixOS/nix,
+  so called authorized committers, sign their commits.
+  - Commits inside merges do not have to be signed as long as the merge commit itself is signed,
+		such that only Committers have to sign their commits.
+
+# Motivation
+[motivation]: #motivation
+
+Signing commits would allow cryptographically verifying that the source code
+hasn't been tampered by an untrusted party.
+
+This would ensure that when pulling from GitHub, or some other source,
+you can check if the source code has been signed by a trusted person.
```

Comment:
I would like to see an example of the kind of attack that this mechanism is supposed to protect against, and an argument that it's something we should actually worry about (relative to other attacks). I think there is significant risk of security theatre here: even if we did all this, it would probably not be too hard for an attacker to get malicious code into nixpkgs, and we would have incurred huge workflow costs for the fantasy of additional security.

Reply:
I've taken your feedback into account.

Yes, this change will not make it impossible to get malicious code into Nixpkgs. As people did with Linux, I could make a patch with hidden vulnerabilities.

How big the workflow overhead is is subjective, but "huge" doesn't seem to match the RFC after the changes I've made. After the transitional phase, the change will for many consist of using GitHub Desktop instead of the Web UI. Others will use the CLI, and some may choose a third solution, depending on what they prefer.

Let's however assume that the workflow costs are enormous.
Currently, we have no way of verifying trust of the commits we update to, other than looking at the channel it came through. If you completely trust GitHub, then yes, the workflow costs are too big for it to be worth it.
However, if you do not trust GitHub, then it is IMO worth it compared to the additional overhead it will add.

If you still feel that using GitHub Desktop is an unacceptable overhead compared to the Web UI, then please do give me feedback on how GitHub Desktop could be made better to match the workflow you get on the web version. Though I have no experience with the codebase, I am willing to fix it for this RFC, as I am the one pushing this forward after all.

Reply:
Is this saying we won't be able to use the GitHub web UI to merge requests even with your change to trusting GitHub's key?

Reply:
No, if we trust GitHub's PGP key then merging through the web UI is fine, since it will use GitHub's PGP key.

As noted in the RFC, we would have a transitional phase, after which we'd have proper tooling, i.e. support in GitHub Desktop, supporting for merging on your mobile, etc. etc., and then we'd be able to remove the trust for GitHub's key without making it hard to merge a PR.

The transitional phase will end when a new RFC is made, then we can discuss the state of the tooling and whether it's possible to do it.

Reply:
> After the transitional phase, the change will for many consist of using GitHub Desktop instead of the Web UI. 

Perhaps we differ about what we consider huge, but IMO asking people to use a tool other than that one they use for ~all of their other OSS work is huge. People will go to merge a PR, think "oh, I need to fire up github desktop", and just... do something else. Small inconveniences can have significant deterrent effects. *Custom* tooling is even worse.

Let me put it another way. At the moment, which is a bigger problem for nixpkgs: time and engagement from committers, or security against attacks like this? I would say the former by a huge margin. So I would not trade the security for a 5% decrease in engagement, especially from key people (several of whom have commented to say that these changes would dissuade them!).

Reply:
I would like to point out that they said that *before* my changes. Of course, they could still be unwilling even with the changes, so let us assume they are still unwilling to use anything that *isn't* the current workflow (though i personally think many would switch to GitHub Desktop *in general*, and use it for all their GitHub activities).

If the extra security is not worth a decrease in engagement, as you say (I personally disagree), then the only solution is to keep the current workflow and somehow make it secure, since we assume that the committers will never get used to any tool that isn't the web UI they currently use.

If we could somehow sign it right from the browser, there would be no problems at all. I think the easiest way of doing this would be having a daemon running on your computer which a browser extension (for example a modified Refined GitHub) connects to, to get it to merge and sign a PR when pressing the merge button on GitHub.

We could also make a NixOS module for it, that handles the daemon side, including cloning the repository, running GPG, etc..

Do you still think there would be a noticeable decrease in engagement with this hypothetical set-up?

Reply:
In any case, the above is a bit off-topic for this RFC. This is relevant for the *next* RFC that comes after this. What we have now in this RFC is simply hypothetical tooling, that will be mentioned in the next RFC. In the case that such tooling is fundamentally impossible, the next RFC just won't happen and the situation will stay as it is for Nixpkgs (even if other repositories adopt the new features).

Reply:
> If the extra security is not worth a decrease in engagement, as you say (I personally disagree), then the only solution is to keep the current workflow and somehow make it secure, since we assume that the committers will never get used to any tool that isn't the web UI they currently use.

We can also simply live with non-optimal security, as many projects do.

> I think the easiest way of doing this would be having a daemon running on your computer which a browser extension

I feel like you've missed the point. The point is not that people desperately want to use a web browser. The point is that people don't want to use a custom, different workflow for nixpkgs, and many will simply choose not to contribute in that case. Running a special daemon and using a special browser extension certainly qualifies as that, IMO!

Reply:
> We can also simply live with non-optimal security, as many projects do.

Yes, and that will happen if the next RFC isn't accepted. This RFC doesn't change any part of the workflow. If there is no way to make Nixpkgs secure, then that's how it will remain, even if other projects start using the features described in this RFC.

> I feel like you've missed the point. The point is not that people desperately want to use a web browser. The point is that people don't want to use a custom, different workflow for nixpkgs, and many will simply choose not to contribute in that case. Running a special daemon and using a special browser extension certainly qualifies as that, IMO!

I'll keep that in mind for the next RFC, thanks.